services/audit/audit.go

```diff
@@ -0,0 +1,211 @@
+	case *repository_model.Repository, *user_model.User, *organization_model.Organization:
+		return typeToDescription(scope)
+	default:
+		panic(fmt.Sprintf("unsupported scope type: %T", s))
+	}
+}
+
+func typeToDescription(val any) TypeDescriptor {
+	switch t := val.(type) {
+	case *repository_model.Repository:
+		return TypeDescriptor{"repository", t.ID, t.FullName(), val}
+	case *user_model.User:
+		if t.IsOrganization() {
+			return TypeDescriptor{"organization", t.ID, t.Name, val}
+		}
+		return TypeDescriptor{"user", t.ID, t.Name, val}
+	case *organization_model.Organization:
+		return TypeDescriptor{"organization", t.ID, t.Name, val}
+	case *user_model.EmailAddress:
+		return TypeDescriptor{"email_address", t.ID, t.Email, val}
+	case *organization_model.Team:
+		return TypeDescriptor{"team", t.ID, t.Name, val}
+	case *auth_model.TwoFactor:
+		return TypeDescriptor{"twofactor", t.ID, "", val}
+	case *auth_model.WebAuthnCredential:
+		return TypeDescriptor{"webauthn", t.ID, t.Name, val}
+	case *user_model.UserOpenID:
+		return TypeDescriptor{"openid", t.ID, t.URI, val}
+	case *auth_model.AccessToken:
+		return TypeDescriptor{"access_token", t.ID, t.Name, val}
+	case *auth_model.OAuth2Application:
+		return TypeDescriptor{"oauth2_application", t.ID, t.Name, val}
+	case *auth_model.OAuth2Grant:
+		return TypeDescriptor{"oauth2_grant", t.ID, "", val}
+	case *auth_model.Source:
+		return TypeDescriptor{"authentication_source", t.ID, t.Name, val}
+	case *user_model.ExternalLoginUser:
+		return TypeDescriptor{"external_account", t.ExternalID, t.ExternalID, val}
+	case *asymkey_model.PublicKey:
+		return TypeDescriptor{"public_key", t.ID, t.Fingerprint, val}
+	case *asymkey_model.GPGKey:
+		return TypeDescriptor{"gpg_key", t.ID, t.KeyID, val}
+	case *secret_model.Secret:
+		return TypeDescriptor{"secret", t.ID, t.Name, val}
+	case *webhook_model.Webhook:
+		return TypeDescriptor{"webhook", t.ID, t.URL, val}
+	case *git_model.ProtectedTag:
+		return TypeDescriptor{"protected_tag", t.ID, t.NamePattern, val}
+	case *git_model.ProtectedBranch:
+		return TypeDescriptor{"protected_branch", t.ID, t.RuleName, val}
+	case *repository_model.PushMirror:
+		return TypeDescriptor{"push_mirror", t.ID, "", val}
+	case *models.RepoTransfer:
+		return TypeDescriptor{"repo_transfer", t.ID, "", val}
+	default:
+		panic(fmt.Sprintf("unsupported type: %T", t))
+	}
+}
+
+func ReleaseReopen() error {
+	var accumulatedErr error
+	for _, a := range appenders {
+		if err := a.ReleaseReopen(); err != nil {
+			if accumulatedErr == nil {
+				accumulatedErr = fmt.Errorf("Error whilst reopening: %w", err)
+			} else {
+				accumulatedErr = fmt.Errorf("Error whilst reopening: %v & %w", err, accumulatedErr)
+			}
+		}
+	}
+	return accumulatedErr
+
+	// TODO Replace with errors.Join > Go 1.20
+	/*var joinedErr error
+	for _, a := range appenders {
+		if err := a.ReleaseReopen(); err != nil {
+			joinedErr = errors.Join(joinedErr, err)
+		}
+	}
+	return joinedErr*/
```

Comment:
Maybe use a build tag for Go 1.19 support, and still supports joining errors, like

```go
//go:build go1.19

package audit

type joinedError struct {
	errs []error
}

func (e *joinedError) Error() string {
	var b string
	for i, err := range e.errs {
		if i > 0 {
			b += ";"
		}
		b += err.Error()
	}
	return b
}

var joinErrors = func(errs ...error) error {
	return &joinedError{errs}
}
```

Reply:
I am sure Gitea is going to use Go 1.20 soon.

Reply:
> Maybe use a build tag for Go 1.19 support, and still supports joining errors, like

Go 1.20 has a new `Unwrap` method which can't be rebuild in 1.19.

Reply:
> I am sure Gitea is going to use Go 1.20 soon.

With current policy of last two supported Go versions, we can only require Go 1.20 when Go 1.21 is out, which is in around 3 months.

Reply:
Now irrelevant with #24859.